Refactoring subscription modal checkmarks to icon format

synicalsyntax opened the pull request:
Turns subscription modal checkmarks into custom icons to fix #10918. Involves a significant amount of CSS refactoring.

Before:
![screenshot at jan 21 16-25-37](https://user-images.githubusercontent.com/15116870/51505125-9a73e880-1d99-11e9-8e47-ec2a6d379e6a.png)

After:
![screenshot at jan 21 16-25-20](https://user-images.githubusercontent.com/15116870/51505127-a069c980-1d99-11e9-9997-9dbe2e26204b.png)


Also includes miscellaneous documentation changes, including custom icon webfont documentation in frontend build processes article and fixing discrepancies in the SVG test tool description.

timabbott commented:
Does this also change the ones in e.g. the "stream description" field?

timabbott commented on static/styles/subscriptions.scss:
```diff
@@ -713,18 +701,42 @@ form#add_new_subscription {
+    display: inline-block;
 }
 
-#subscription_overlay .stream-description .stream-description-editable:empty::after,
```
This line seems to have been split with a comma?  How did you generate this commit ?  I would have imagined one would have done it with just copy-pasting rearranging things to avoid risk of introducing unexpected changes like this.

```
#subscription_overlay .stream-description,
￼	.stream-description-editable:empty::after {
￼```	

If you're doing more complex changes to some of the blocks to re-split them or something, please do that in a separate commit from simple reordering -- it's really unpleasant to verify a reordering commit squashed with other changes.

synicalsyntax commented:
@timabbott Added a separate commit for splitting out the selectors like you asked, please review!

timabbott commented:
I merged the first commit.  For ` subs: Split stream row selectors for SASS nested selector feature.`, it seems like you lost this block:
```
-.stream-row::-moz-selection,
-.stream-row .icon .hashtag::-moz-selection {
+.stream-row::-moz-selection {
     background-color: transparent;
 }
 
-.stream-row::selection,
-.stream-row .icon .hashtag::selection {
+.stream-row::selection {
     background-color: transparent;
 }
```

(At least, that change isn't a no-op)

timabbott commented on static/styles/subscriptions.scss:
```diff
@@ -590,167 +590,171 @@ form#add_new_subscription {
+            margin: 0% 15%;
+        }
+
+        &.checked {
```
I think at least this use of  `&` is buggy, see https://css-tricks.com/the-sass-ampersand/.  The original rule here was 
`.stream-row .checked svg`, not `.stream-row.checked svg` which is what you have here...

synicalsyntax replied:
No, I believe I didn't make a mistake here. The original CSS was:
```css
.stream-row .check svg {
    fill: transparent;
    width: 70%;
    margin: 0% 15%;
}

.stream-row .checked svg {
    fill: hsl(170, 48%, 54%);
}

.stream-row .check.checked:hover svg {
    opacity: 0.5;
}
```

The DOM looks like:
```html
<div class="stream-row" data-stream-id="1" data-stream-name="announce">
    <div class="check checked sub_unsub_button">
```
So, really, `.stream-row .checked svg` should be `.stream-row .check.checked svg`. Also, `&` refers to the parent selector of the current element, which is `.check` and not `.stream-row` since it's nested under the `.check` selector which is nested under `.stream-row`.

timabbott replied:
Yes, but if I'm not mistaken, `&.checked` is how you generate `.stream-row.checked`, not `.stream-row .checked` - i.e. expecting the checked to be on the same element as the stream-row.

synicalsyntax commented:
@timabbott Actually, I split that block out and moved it up:
![screenshot at feb 18 16-57-08](https://user-images.githubusercontent.com/15116870/52983101-4bbc6d00-339e-11e9-9871-8eeb2f776b97.png)

timabbott commented:
@synicalsyntax did you mean to close this?  If you don't have time for it, maybe @YJDave can pick it up.
